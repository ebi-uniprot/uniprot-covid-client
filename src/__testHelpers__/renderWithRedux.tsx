/* eslint-disable */
import React from 'react';
import { createStore, applyMiddleware } from 'redux';
import { render } from '@testing-library/react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { createMemoryHistory, MemoryHistory } from 'history';
import rootReducer from '../state/rootReducer';
import { Router, Route } from 'react-router-dom';

type RenderOptions = {
  route?: string;
  history?: MemoryHistory<any>;
  path?: string;
  initialState?: any;
  store?: any;
};

const renderWithRedux = (
  ui: React.Component,
  {
    route = '',
    history = createMemoryHistory({ initialEntries: [route] }),
    path,
    initialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
  }: RenderOptions = {}
) => {
  return {
    ...render(
      <Provider store={store}>
        <Router history={history}>
          {path ? <Route path={path} render={() => ui} /> : ui}
        </Router>
      </Provider>
    ),
    store,
    history,
  };
};

export default renderWithRedux;
