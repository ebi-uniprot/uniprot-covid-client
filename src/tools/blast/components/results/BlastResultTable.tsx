/* eslint-disable camelcase */
import React, { FC, useCallback, useState, useRef, useMemo } from 'react';
import { DataTable, DENSITY_COMPACT, Chip, Loader } from 'franklin-sites';
import { Link } from 'react-router-dom';
import ProtvistaTrack from 'protvista-track';
import ProtvistaNavigation from 'protvista-navigation';

import { ReviewedUnreviewed } from '../../../../uniprotkb/components/protein-data-views/UniProtKBTitle';

import useStaggeredRenderingHelper from '../../../../shared/hooks/useStaggeredRenderingHelper';

import { loadWebComponent } from '../../../../shared/utils/utils';

import { BlastResults, BlastHsp, BlastHit } from '../../types/blastResults';
import { HSPDetailPanelProps } from './HSPDetailPanel';
import {
  EntryType,
  UniProtkbAPIModel,
} from '../../../../uniprotkb/adapters/uniProtkbConverter';

import './styles/BlastResultTable.scss';
import { EnrichedBlastHit } from './BlastResult';

const BlastSummaryTrack: FC<{
  hsp: BlastHsp;
  extra?: UniProtkbAPIModel;
  queryLength: number;
  hitLength: number;
  setHspDetailPanel: (props: HSPDetailPanelProps) => void;
  hitAccession: string;
}> = ({
  hsp,
  queryLength,
  hitLength,
  setHspDetailPanel,
  hitAccession,
  extra,
}) => {
  const {
    hsp_query_from,
    hsp_query_to,
    hsp_identity,
    hsp_bit_score,
    hsp_expect,
  } = hsp;

  const setTrackData = useCallback(
    (node): void => {
      if (node) {
        /**
         * TODO - would be nice to add gaps
         * at some point
         */
        // eslint-disable-next-line no-param-reassign
        node.data = [
          {
            start: 1,
            end: hsp_query_from,
            shape: 'line',
            color: '#014371',
            opacity: hsp_identity / 100,
          },
          {
            start: hsp_query_from,
            end: hsp_query_to,
            // franklin $colour-sapphire-blue
            color: '#014371',
            opacity: hsp_identity / 100,
          },
          {
            start: hsp_query_to,
            end: hitLength > hsp_query_to ? hitLength : hsp_query_to,
            shape: 'line',
            color: '#014371',
            opacity: hsp_identity / 100,
          },
        ];
      }
    },
    [hsp_query_from, hsp_query_to, hsp_identity, hitLength]
  );

  return (
    <div className="data-table__blast-hsp__tracks">
      <section className="data-table__blast-hsp__blast-track">
        <protvista-track
          data-testid="blast-summary-track"
          length={queryLength}
          height={10}
          ref={setTrackData}
          title={`Start: ${hsp_query_from}\nEnd: ${hsp_query_to}\nHit Length: ${hitLength}`}
          onClick={() => {
            setHspDetailPanel({
              hsp,
              hitAccession,
              onClose: () => null,
              queryLength,
              hitLength,
              extra,
            });
          }}
        />
      </section>
      <span className="data-table__blast-hsp__blast-params">
        <Chip compact title="Identity">{`${hsp_identity}%`}</Chip>
        <Chip compact title="Score">{`${hsp_bit_score}`}</Chip>
        <Chip compact title="e-value">{`${hsp_expect}`}</Chip>
      </span>
    </div>
  );
};

const BlastSummaryHsps: FC<{
  hsps: BlastHsp[];
  queryLength: number;
  hitLength: number;
  hitAccession: string;
  extra?: UniProtkbAPIModel;
  setHspDetailPanel: (props: HSPDetailPanelProps) => void;
}> = ({
  hsps,
  queryLength,
  hitLength,
  setHspDetailPanel,
  hitAccession,
  extra,
}) => {
  const [collapsed, setCollapsed] = useState(true);

  const hspsOrderedByScore = hsps.sort(
    (hspA, hspB) => hspB.hsp_bit_score - hspA.hsp_bit_score
  );

  return (
    <div className="data-table__blast-hsp">
      <div>
        <BlastSummaryTrack
          hsp={hspsOrderedByScore[0]}
          queryLength={queryLength}
          hitLength={hitLength}
          hitAccession={hitAccession}
          extra={extra}
          setHspDetailPanel={setHspDetailPanel}
        />
        {hspsOrderedByScore.length > 1 &&
          !collapsed &&
          hspsOrderedByScore
            .slice(1)
            .map((hsp) => (
              <BlastSummaryTrack
                hsp={hsp}
                queryLength={queryLength}
                hitLength={hitLength}
                key={`hsp_${hsp.hsp_num}`}
                hitAccession={hitAccession}
                extra={extra}
                setHspDetailPanel={setHspDetailPanel}
              />
            ))}
      </div>
      {hspsOrderedByScore.length > 1 && collapsed && (
        <button type="button" onClick={() => setCollapsed(false)}>{`+${
          hspsOrderedByScore.length - 1
        } more`}</button>
      )}
    </div>
  );
};

const BlastResultTable: FC<{
  data: BlastResults | null;
  selectedEntries: string[];
  handleSelectedEntries: (rowId: string) => void;
  setHspDetailPanel: (props: HSPDetailPanelProps) => void;
  loading: boolean;
}> = ({
  data,
  selectedEntries,
  handleSelectedEntries,
  setHspDetailPanel,
  loading,
}) => {
  // logic to keep stale data available
  const hitsRef = useRef<BlastHit[]>([]);
  if (data?.hits.length || !loading) {
    hitsRef.current = data?.hits || [];
  }

  const nItemsToRender = useStaggeredRenderingHelper(
    hitsRef.current.length
      ? { first: 25, max: hitsRef.current.length }
      : undefined
  );

  // The "query" column header
  const queryColumnHeaderRef = useCallback(
    (node) => {
      if (node && data) {
        const { query_len } = data;
        // eslint-disable-next-line no-param-reassign
        node.data = [
          {
            start: 1,
            end: query_len,
          },
        ];
      }
    },
    [data]
  );

  loadWebComponent('protvista-track', ProtvistaTrack);
  loadWebComponent('protvista-navigation', ProtvistaNavigation);

  const queryLen = data?.query_len;
  const columns = useMemo(() => {
    if (queryLen === undefined) {
      return [];
    }
    return [
      {
        label: 'Accession',
        name: 'accession',
        render: ({ hit_acc, hit_db }: BlastHit) => {
          let reviewImg;
          switch (hit_db) {
            case 'SP':
              reviewImg = <ReviewedUnreviewed entryType={EntryType.REVIEWED} />;
              break;
            case 'TR':
              reviewImg = (
                <ReviewedUnreviewed entryType={EntryType.UNREVIEWED} />
              );
              break;
            default:
              reviewImg = null;
          }
          return (
            <Link to={`/uniprotkb/${hit_acc}`}>
              {reviewImg}
              {hit_acc}
            </Link>
          );
        },
        width: '8rem',
      },
      {
        label: 'Gene',
        name: 'gene',
        render: ({ hit_uni_gn }: BlastHit) => hit_uni_gn,
        width: '5rem',
      },
      {
        label: 'Protein',
        name: 'protein_name',
        render: ({ hit_uni_de }: BlastHit) => hit_uni_de,
        ellipsis: true,
      },
      {
        label: 'Organism',
        name: 'organism',
        render: ({ hit_uni_ox, hit_uni_os }: BlastHit) => (
          <Link to={`/taxonomy/${hit_uni_ox}`}>{hit_uni_os}</Link>
        ),
        ellipsis: true,
      },
      {
        label: (
          <div className="query-sequence-wrapper">
            <protvista-navigation
              ref={queryColumnHeaderRef}
              length={queryLen}
              height={10}
              title="Query"
            />
          </div>
        ),
        name: 'alignment',
        width: '40vw',
        render: ({ hit_hsps, hit_len, hit_acc, extra }: EnrichedBlastHit) => (
          <BlastSummaryHsps
            hsps={hit_hsps}
            queryLength={queryLen}
            hitLength={hit_len}
            hitAccession={hit_acc}
            setHspDetailPanel={setHspDetailPanel}
            extra={extra}
          />
        ),
      },
    ];
  }, [queryLen, queryColumnHeaderRef, setHspDetailPanel]);

  if (loading && !hitsRef.current.length) {
    return <Loader />;
  }

  if (!data) {
    return null;
  }

  return (
    <div className={loading ? 'loading-data-table' : undefined}>
      <DataTable
        getIdKey={({ hit_acc }: { hit_acc: string }) => hit_acc}
        density={DENSITY_COMPACT}
        columns={columns}
        data={hitsRef.current.slice(0, nItemsToRender)}
        selectable
        selected={selectedEntries}
        onSelect={handleSelectedEntries}
        fixedLayout
      />
    </div>
  );
};

export default BlastResultTable;
