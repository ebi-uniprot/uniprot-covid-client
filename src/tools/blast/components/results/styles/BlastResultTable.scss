.data-table {
  .query-sequence-wrapper {
    width: 32vw;
    margin-left: 0.2rem;
    // Hide unwanted items from navigation
    protvista-navigation {
      .zoom-polygon,
      .brush,
      .start-label,
      .end-label {
        display: none;
      }
    }
  }
}

.data-table__blast-hsp {
  $self: &;
  display: flex;
  .chip {
    margin-top: 0;
  }
  &__tracks {
    position: relative;
    #{ $self }__blast-track {
      /**
      * Magic number preventing the forever growing
      * protvista-track
      */
      width: 32vw;
      cursor: pointer;
      &:hover {
        background: #d2dce3;
      }
      svg {
        margin: 0.2rem;
      }
    }
    #{ $self }__blast-params {
      position: absolute;
      top: -0.6rem;
      right: 0.1rem;
    }
  }
  button {
    margin-left: 0.5rem;
  }
}

@keyframes loading {
  from {
    opacity: 1;
  }
  to {
    opacity: 0.5;
  }
}

.loading-data-table .data-table {
  animation: alternate loading 1000ms infinite ease-in-out;
}
