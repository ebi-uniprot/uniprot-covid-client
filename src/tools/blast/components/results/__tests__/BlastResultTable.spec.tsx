import React from 'react';
import BlastResultTable from '../BlastResultTable';
import blastResultsMockData from '../../../../__mocks__/server-jobs/example-truncated.json';
import renderWithRedux from '../../../../../shared/__test-helpers__/RenderWithRedux';
import { fireEvent } from '@testing-library/react';

let component;

describe('BlastResultTable tests', () => {
  beforeEach(() => {
    component = renderWithRedux(
      <BlastResultTable data={blastResultsMockData} />
    );
  });

  it('should render the results table', () => {
    const { asFragment } = component;
    expect(asFragment()).toMatchSnapshot();
  });

  it('should toggle the extra hss', async () => {
    const { getByText, findAllByTestId } = component;
    let hspTracks = await findAllByTestId('blast-summary-track');
    expect(hspTracks.length).toBe(1);
    const toggle = getByText('+1 more');
    fireEvent.click(toggle);
    hspTracks = await findAllByTestId('blast-summary-track');
    expect(hspTracks.length).toBe(2);
  });
});
