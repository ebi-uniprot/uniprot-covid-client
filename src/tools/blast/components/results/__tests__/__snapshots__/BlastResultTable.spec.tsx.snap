// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlastResultTable tests should render the results table 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data-loader__wrapper"
    >
      <div
        class="data-loader__scroll-container"
      >
        <table
          class="data-table data-table--compact  data-table--fixed"
        >
          <thead
            class="data-table__header"
          >
            <tr
              class="data-table__row"
            >
              <th
                class="data-table__header-cell data-table__header-cell--checkbox"
              >
                 
              </th>
              <th
                class="data-table__header-cell "
                style="width: 8rem;"
              >
                Accession
              </th>
              <th
                class="data-table__header-cell "
                style="width: 5rem;"
              >
                Gene
              </th>
              <th
                class="data-table__header-cell "
                style="width: auto;"
              >
                Protein
              </th>
              <th
                class="data-table__header-cell "
                style="width: auto;"
              >
                Organism
              </th>
              <th
                class="data-table__header-cell "
                style="width: 40vw;"
              >
                <div
                  class="query-sequence-wrapper"
                >
                  <protvista-navigation
                    height="10"
                    length="30"
                    title="Query"
                  />
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class="data-table__table__body"
          >
            <tr>
              <td
                class="data-table__cell "
              >
                <input
                  aria-label="U6MKN0"
                  type="checkbox"
                />
              </td>
              <td
                class="data-table__cell  data-table__cell--ellipsis"
              >
                <a
                  href="/uniprotkb/U6MKN0"
                >
                  <span
                    class="uniprot-title__status icon--unreviewed"
                    title="This marks an unreviewed entry"
                  >
                    <svg
                      fill="currentColor"
                      viewBox="0 0 400 512"
                    >
                      <path
                        d="M400 144H256V0l144 144zm-176 32V0H0v512h400V176H224z"
                      />
                    </svg>
                  </span>
                  U6MKN0
                </a>
              </td>
              <td
                class="data-table__cell  data-table__cell--ellipsis"
              >
                ENH_00046930
              </td>
              <td
                class="data-table__cell  data-table__cell--ellipsis"
              >
                Uncharacterized protein
              </td>
              <td
                class="data-table__cell  data-table__cell--ellipsis"
              >
                <a
                  href="/taxonomy/51315"
                >
                  Eimeria necatrix
                </a>
              </td>
              <td
                class="data-table__cell  data-table__cell--ellipsis"
              >
                <div
                  class="data-table__blast-hsp"
                >
                  <div>
                    <div
                      class="data-table__blast-hsp__tracks"
                    >
                      <section
                        class="data-table__blast-hsp__blast-track"
                      >
                        <protvista-track
                          data-testid="blast-summary-track"
                          height="10"
                          length="30"
                          title="Start: 3
End: 30
Hit Length: 486"
                        />
                      </section>
                      <span
                        class="data-table__blast-hsp__blast-params"
                      >
                        <span
                          class="chip chip--compact"
                          title="Identity"
                        >
                          7.1%
                        </span>
                        <span
                          class="chip chip--compact"
                          title="Score"
                        >
                          41.5874
                        </span>
                        <span
                          class="chip chip--compact"
                          title="e-value"
                        >
                          0.0048
                        </span>
                      </span>
                    </div>
                  </div>
                  <button
                    type="button"
                  >
                    +1 more
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
